File: 278. First Bad Version/main.rs
isBadVersion(version:i32)-> bool;

// The API isBadVersion is defined for you.
// isBadVersion(version:i32)-> bool;
// to call it use self.isBadVersion(version)

impl Solution {
    pub fn first_bad_version(&self, n: i32) -> i32 {
        let mut left = 1;
        let mut right = n;
        let mut first = -1;

        while left <= right {
            let mid = left + (right - left) / 2;
            if self.isBadVersion(mid) {
                first = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return first;
    }
}

File: 278. First Bad Version/main.cpp
bool isBadVersion(int version);

// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int left = 1;
        int right = n;
        int first = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (isBadVersion(mid)) {
                first = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return first;
    }
};